include/BareCpper/ATsamd/ExternalInterruptController_ATsamd5x.hpp

```diff
@@ -0,0 +1,182 @@
+#ifndef EXTERNAL_INTERRUPT_CONTROLLER_SAMD51_HPP
+#define EXTERNAL_INTERRUPT_CONTROLLER_SAMD51_HPP
+
+#if !__SAMD51__
+#  error "SAMD library error, please check and update accordingly."
+#endif
+
+#include <functional>
+
+#include "../Gpio.hpp"
+#include <sam.h>
+
+namespace BareCpper
+{
+  namespace SAMD51
+  {
+    /**
+     * Edge or level configuration for
+     * the external interrupt/event
+     * 
+     */
+    enum class ExternalInterruptSense : uint8_t
+    {
+      None  = 0x0,
+      Rise  = 0x1,
+      Fall  = 0x2,
+      Both  = 0x3,
+      High  = 0x4,
+      Low   = 0x5
+    };
+
+    enum class ExternalInterruptDebouncer : uint8_t
+    {
+      Off = 0xFF,
+      Prescaler2    = 0x0,
+      Prescaler4    = 0x1,
+      Prescaler8    = 0x2,
+      Prescaler16   = 0x3,
+      Prescaler32   = 0x4,
+      Prescaler64   = 0x5,
+      Prescaler128  = 0x6,
+      Prescaler256  = 0x7,
+    };
+
+    class ExternalInterruptController
+    {
+      public:
+        static constexpr uint8_t NumberChannels = 16;
+        /**
+         * Initialize the EIC
+         * 
+         */
+        void init()
+        {
+          // EIC MCLK is enabled by default
+          // Use GCLK1 for EIC (max is 100 MHz)
+          GCLK->PCHCTRL[EIC_GCLK_ID].reg = GCLK_PCHCTRL_GEN_GCLK1_Val | (1 << GCLK_PCHCTRL_CHEN_Pos);
+          // Wait for clock synchronization
+				  while (GCLK->SYNCBUSY.reg);
+          // Reset the EIC
+          reset();
+          // Enable the EIC
+          enable();
+        }
+
+        using CallbackT = std::function<void()>;
+        
+        template<typename PinT>
+        void attachInterrupt(const PinT& pin
+                            , const CallbackT& callback
+                            , const ExternalInterruptSense& mode
+                            , const ExternalInterruptDebouncer& debounce = ExternalInterruptDebouncer::Off)
+        {
+          // registers are enable-protected, disable EIC before configuring if it is enabled
+          const bool shouldRestart = EIC->CTRLA.reg & EIC_CTRLA_ENABLE;
+          if(shouldRestart) disable();
+          // set the Pin alternative function to peripheral A (EIC)
+          BareCpper::gpioFunction<PinT>(pin, std::optional<ATsamd5x::Peripheral>{ATsamd5x::Peripheral::A});
+          // configure the EIC sense mode
+          constexpr auto channel = (PinT::id().pin > 15) ? (PinT::id().pin - 16) : PinT::id().pin;
+          constexpr uint8_t configRegNum = (channel > 7);
+          constexpr uint8_t configRegShift = (channel - 8*configRegNum) << 2; //< channel config starts at bit 4^channel
+          EIC->CONFIG[configRegNum].reg &= ~(0x7 << configRegShift);
```

Comment:
Minor-Optimize: beneficial to reduce the number of Read-write operations to 'Special' registers. i.e. `EIC->CONFIG[configRegNum].reg`  could be read one and written once

Reply:
Other thought for future optimising, The 'register' values can be calculated before setting i.e.:
```
Read current..
calculate new...
disable/unlock
apply ++ relock/enable in one
```

Reply:
agreed